### ISSUE TITLE
Create OckamD in Rust

### ISSUE THREAD

**Issue by mrinalwadhwa** [ockam-network/ockam#412]
Ockam Daemon / OckamD is a daemon program written in rust and distributed as a binary for major operating systems / architectures. (we can start with just linux/macos support and add other operating systems later)

* It opens up raw input/output portals for various protocols.
* A raw input portal - turns a message that is received on the input into an ockam routing protocol message.
* The first one we're interested in supporting is an input portal for STDIN that turn every line that is input on stdin into an ockam routing protocol message.

We want this to be compatible with InfluxDB External Plugin
https://github.com/influxdata/telegraf/blob/master/docs/EXTERNAL_PLUGINS.md

Steps:
- [x] Create a crate called daemon in https://github.com/ockam-network/ockam/tree/develop/implementations/rust where the code of ockam daemon will live.
- [x] Add a binary program that  will depend on the `ockam` crate https://github.com/ockam-network/ockam/tree/develop/implementations/rust/ockam
- [x] Add some means to define command line arguments
- [x] Instantiate the Ockam Software Vault via persisted FilesystemVault
- [x] Use routing and secure channels form the `ockam` crate to setup a secure channel with another Ockam node.

**Comment by mrinalwadhwa** [comment]
The CLI and Daemon may be the same binary program that can be started in different modes. (interactive or not) #326

**Comment by nilslice** [comment]
@mrinalwadhwa - any objection to using an async runtime in `ockamd`? Since `telegraf ` is written in Go, I don't think `ockamd` for `telegraf` would run in an environment where `tokio` or another Rust runtime would be off-limits from a resource standpoint... but let me know if you are already aware of cases which that would not be true. 

I was thinking that when `ockamd` is run in interactive-mode, it would send messages over TCP or HTTP via localhost to the `ockamd` running in server-mode. Since `telegraf` will have spawned `ockamd` in server-mode (using STDIN for passing Influx messages) I think it will be more difficult to try and re-use STDIN for interactive-mode -> server-mode process communication. `telegraf` will have a handle to fd/0 on the `ockamd` server-mode process, but `ockamd` interactive-mode would need to look it up each time the binary is run, requiring a different implementation for various OSes. 

Instead of TCP or HTTP, other options I've considered for interactive-mode -> server-mode processes include:
- named pipes (likely also to require OS-specific implementations)
- file on disk
- custom IPC (still needs to keep track of the PID or OS-specific process handle)
- setting/reading ENV variables (need to consider OS-scoping of environment vars)

I think TCP or HTTP have the least downside compared to each of the others. The caveat is the need to reserve a port for this as a default, or have the `telegraf` config file include one in an extra parameter: 
```toml
# /etc/telegraf/telegraf.conf
[[inputs.execd]]
  command = ["ockamd", "-s", "-p", "11199"]
```

However, in the case that the config file has an override for the server-mode port (e.g. `11199` above), then the interactive-mode invocations would need to also use the same in an extra parameter: `$ ockamd -i -p 11199 ...` 

My opinion is that TCP (using serializable messages) and a sane default, but configurable, port is the best option with the least downside. If you had other options in mind, please share.

**Comment by nilslice** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by nilslice** [comment]
recheckcla

**Comment by mrinalwadhwa** [comment]
Tokio or something similar sounds like a great idea 👍 
Could you post some quick research / comparison for what the various choices are.

Sending messages to ockamd over a network protocol makes sense.
UDP over the loopback address may be easiest since we're already implementing a UDP ockam transport 
> My opinion is that TCP (using serializable messages) and a sane default, but configurable, port is the best option with the least downside.

**Comment by nilslice** [comment]
> Could you post some quick research / comparison for what the various choices are.

There is surprisingly sparse comparison available publicly, but here are the options I would consider:
https://tokio.rs/
https://github.com/stjepang/smol
https://async.rs/

`tokio` is the most broadly used in the community, and would integrate with more async crates (not that we need to use many if any at all). There are ways to make the other two compatible with `tokio` eco-system crates though. I don't have much experience with the others in a production environment either.

Another consideration re: async is if `ockamd` will need to run in an embedded environment. If so, there is [work to support async ](https://ferrous-systems.com/blog/stable-async-on-embedded/) in `no_std` environments, but it would take more research to come to a conclusion there. If we know it will need to run within the constraints of `no_std`, let's decide if that is something to defer until later or support now.

+1 to UDP transport, I will take a look at the existing ockam support.

**Comment by mrinalwadhwa** [comment]
`ockamd` with `no_std` is not a requirement for now ... we can assume `ockamd` is running on a machine that is running Linux.

**Comment by mrinalwadhwa** [comment]
@mikelodder7 could you share your thoughts / advice on which async lib we should use for `ockamd`

**Comment by mikelodder7** [comment]
async.rs is my preference since its more rust like than tokio.

**Comment by nilslice** [comment]
Documentation card reminder: https://github.com/orgs/ockam-network/projects/13#card-46413205

**Comment by nilslice** [comment]
Re-opening (my fault the link auto-closed this in midpoint PR #436)
